File: frontend/src/api/health.ts
export type SubsystemCheck = {
  name: string;
  passed: boolean;
  detail: string | null;
};

export type ReadinessReport = {
  status: string;
  checked_at: string;
  checks: SubsystemCheck[];
};

const BASE: string = import.meta.env.VITE_API_BASE_URL ?? "";

export async function fetchHealth(): Promise<ReadinessReport> {
  const res = await fetch(`${BASE}/health`);
  // /health returns 200 (healthy) or 503 (not-healthy); both carry the report body.
  if (res.status !== 200 && res.status !== 503) {
    throw new Error(`unexpected status ${res.status}`);
  }
  return (await res.json()) as ReadinessReport;
}

File: backend/app/domain/health.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from enum import StrEnum


class HealthStatus(StrEnum):
    HEALTHY = "healthy"
    NOT_HEALTHY = "not_healthy"


class SubsystemName(StrEnum):
    DATA_STORE = "data_store"
    MESSAGE_BROKER = "message_broker"
    WORKER_POOL_CPU = "worker_pool_cpu"
    WORKER_POOL_IO = "worker_pool_io"


@dataclass(frozen=True, slots=True)
class SubsystemCheck:
    """Result of one shallow connectivity check inside the aggregate readiness report."""

    name: SubsystemName
    passed: bool
    detail: str | None = None


@dataclass(frozen=True, slots=True)
class ReadinessReport:
    """Aggregate verdict returned by GET /health (binary; no degraded state in this slice)."""

    checks: tuple[SubsystemCheck, ...]
    checked_at: datetime = field(default_factory=lambda: datetime.now(UTC))

    @property
    def status(self) -> HealthStatus:
        return HealthStatus.HEALTHY if self.healthy else HealthStatus.NOT_HEALTHY

    @property
    def healthy(self) -> bool:
        return all(check.passed for check in self.checks)
